src/nfa/intersection.cc

```diff
@@ -16,177 +16,254 @@
-    }
-}
+    auto add_product_symbol_post = [&product,&get_product_state](const State lhs_source, const State rhs_source, SymbolPost& new_product_symbol_post)
+    {
+        if (new_product_symbol_post.empty()) { return; }
+
+        State product_source = get_product_state(lhs_source,rhs_source);
+
+        StatePost &product_state_post{product.delta.mutable_state_post(product_source)};
+        if (product_state_post.empty() || new_product_symbol_post.symbol > product_state_post.back().symbol) {
+            product_state_post.push_back(std::move(new_product_symbol_post));
+        }
+        else {
+            //this case happens when inserting epsilon transitions
+            auto symbol_post_it = product_state_post.find(new_product_symbol_post.symbol);
+            if (symbol_post_it == product_state_post.end()) {
+                product_state_post.insert(std::move(new_product_symbol_post));
+            }
+            else {
+                symbol_post_it->insert(new_product_symbol_post.targets);
+            }
+        }
+    };
 
 /**
  * Create product state and its transitions.
- * @param[out] product Created product automaton.
- * @param[out] product_map Created product map.
- * @param[out] pairs_to_process Set of product states to process
- * @param[in] lhs_state_to Target state in NFA @c lhs.
- * @param[in] rhs_state_to Target state in NFA @c rhs.
- * @param[out] intersect_transitions Transitions of the product state.
+ * @param[in] lhs_target Target state in NFA @c lhs.
+ * @param[in] rhs_target Target state in NFA @c rhs.
+ * @param[out] product_symbol_post Transitions of the product state.
  */
-void create_product_state_and_trans(
-            Nfa& product,
-            std::unordered_map<std::pair<State,State>, State>& product_map,
-            const Nfa& lhs,
-            const Nfa& rhs,
-            std::deque<std::pair<State,State>>& pairs_to_process,
-            const State lhs_state_to,
-            const State rhs_state_to,
-            SymbolPost& intersect_transitions
-) {
-    const std::pair<State,State> intersect_state_pair_to(lhs_state_to, rhs_state_to);
-    State intersect_state_to;
-    if (product_map.find(intersect_state_pair_to) == product_map.end()) {
-        intersect_state_to = product.add_state();
-        product_map[intersect_state_pair_to] = intersect_state_to;
-        pairs_to_process.push_back(intersect_state_pair_to);
-
-        if (lhs.final[lhs_state_to] && rhs.final[rhs_state_to]) {
-            product.final.insert(intersect_state_to);
-        }
-    } else {
-        intersect_state_to = product_map[intersect_state_pair_to];
-    }
-    intersect_transitions.insert(intersect_state_to);
-}
-
-} // Anonymous namespace.
+    auto create_product_state_and_move = [&product, &pairs_to_process, &product_contains, &get_product_state, &insert_product_state, &final_condition]
+            (const State lhs_target, const State rhs_target, SymbolPost& product_symbol_post)
+    {
```

Comment:
Simplify lambda capture
```suggestion
    auto create_product_state_and_move = [&](const State lhs_target, const State rhs_target, SymbolPost& product_symbol_post) {
```

Reply:
he, you can do that?

Reply:
It is an automatic reference capture: everything needed inside the lambda that you have not specified earlier with another kind of capture will be captured with this kind of capture (mutable reference in this case).